### PR TITLE
SPARC assembly: Don't file aes-cbc on T4 with small sizes.

### PR DESCRIPTION
**Pull request by sebastianas**
The "openssl speed -testmode -seconds 1 -bytes 1 aes-128-cbc" test revealed that the assembly code is crashing if length is less than 16. The code shifts the provided length by 4 and than subtracts one until the length hits zero. If it was already zero then it underflows the counter and continues until it segfaults on reading or writing.

Replace the check against 0 with less than 15.

Signed-off-by: Sebastian Andrzej Siewior <sebastian@breakpoint.cc>

### CODE REVIEW

**Comment by sebastianas** [comment]
I am not sure API wise if this can also happen to the other versions such as xts or ctr which is also in this file. With this, the testsuite passes on "UltraSparc T4 (Niagara4)"

**Comment by Sashan** [comment]
Thank you for pull request/bug report. would you be able to try patch below:
```
diff --git a/crypto/perlasm/sparcv9_modes.pl b/crypto/perlasm/sparcv9_modes.pl
index 76a2727aba..355ea007d4 100644
--- a/crypto/perlasm/sparcv9_modes.pl
+++ b/crypto/perlasm/sparcv9_modes.pl
@@ -116,12 +116,14 @@ $::code.=<<___;
 	add		$inp, 16, $inp
 
 	brnz,pn		$ooff, 2f
-	sub		$len, 1, $len
 
 	std		%f0, [$out + 0]
 	std		%f2, [$out + 8]
-	brnz,pt		$len, .L${bits}_cbc_enc_loop
 	add		$out, 16, $out
+	bnz,pt		$len, .L${bits}_cbc_enc_loop_done
+	sub		$len, 1, $len
+	ba,a		.L${bits}_cbc_enc_loop
+.L${bits}_cbc_enc_loop_done:
 ___
 $::code.=<<___ if ($::evp);
 	st		%f0, [$ivec + 0]
@@ -338,12 +340,14 @@ $::code.=<<___;
 	movxtod		%o1, %f14
 
 	brnz,pn		$ooff, 2f
-	sub		$len, 1, $len
-
 	std		%f0, [$out + 0]
 	std		%f2, [$out + 8]
-	brnz,pt		$len, .L${bits}_cbc_dec_loop2x
 	add		$out, 16, $out
+	brz,pn		$len, .L{$bits}_cbc_dec_loop_done
+	sub		$len, 1, $len
+
+	ba,a		.L${bits}_cbc_dec_loop
+.L${bits}_cbc_dec_loop_done:
 ___
 $::code.=<<___ if ($::evp);
 	st		%f12, [$ivec + 0]
```

The idea is to check for 0 length later so the function should be able to decode/encode input which is shorter than 16 bytes. I don't have access to sparc so I could not try that.

thanks.

**Comment by paulidale** [comment]
CBC mode is block based.  It doesn't make sense for sizes not a multiple of 16.

**Review by Sashan — APPROVED** [review]
Seeing comment from @paulidale the change looks good.
I have no further comments/suggestions.
thanks.

**Comment by sebastianas** [comment]
On 2024-10-08 20:09:00 [-0700], Pauli wrote:
> CBC mode is block based.  It doesn't make sense for sizes not a multiple of 16.

The "openssl speed" can do it and it is part of the test suite. I don't
know what other implementations do but if it shouldn't be fixed in sparc
then it should be forbidden at the API level (if possible).

I don't disagree that it does not make sense to use length which is a
multiple of 16.

Sebastian

**Comment by sebastianas** [comment]
On 2024-10-09 11:16:46 [-0700], Tomáš Mráz wrote:
> > The "openssl speed" can do it and it is part of the test suite.
> 
> The implication is that openssl speed is wrong - it should report an
> error if the length is not a multiply of a block size and it must not
> call EVP_Cipher with such length.

But it is the official API an can be called by everyone. Or do you want
tell everyone that a crash here is okay because length should be a
multiple of block size?

The zero check is already there…

Sebastian

**Comment by paulidale** [comment]
The documentation clearly states this is undefined behaviour, so we have told everyone about it.

OpenSSL has many APIs that will crash or do random things if passed "bad" arguments.  We're ever so slowly cleaning these up and new code generally doesn't make such assumptions.

We should clean this up at some point.

**Comment by t8m** [comment]
1. openssl speed should be fixed

2. The question is what is the right behavior for such invalid calls. I would argue that crash might be safer than returning an error as an application that does not check for error return might expect that the encryption has happened and in case in-place encryption is being done it would happily send out the plaintext.

Yes, it is good that for new calls we properly define the semantics for all input arguments. But for existing calls it can be actually dangerous to define previously undefined behavior even if it means just returning an error.

**Comment by sebastianas** [comment]
On 2024-10-10 02:41:15 [-0700], Tomáš Mráz wrote:
> 1. openssl speed should be fixed

align up to block or a different API?

> 2. The question is what is the right behavior for such invalid calls.
> I would argue that crash might be safer than returning an error as an
> application that does not check for error return might expect that the
> encryption has happened and in case in-place encryption is being done
> it would happily send out the plaintext.

The sparc code has already a length check and I extended it to cover the
< 16 bytes case.

If you prefer a crash for a length which is a multiple of the block size
then I would suggest to add something with the API that we crash
everywhere and not just on sparc. I've uploaded beta1 to Debian
experimental and it build and run the testuite almost everywhere (riscv,
alpha and hurd-amd64 are still pending). Only Sparc failed so far.

> Yes, it is good that for new calls we properly define the semantics
> for all input arguments. But for existing calls it can be actually
> dangerous to define previously undefined behavior even if it means
> just returning an error.

Sebastian

**Comment by sebastianas** [comment]
Could this be applied, please?
3.4.0 still crashes/ fails here in the testsuite.

**Comment by t8m** [comment]
I've created PR #25792 to fix the crashing speed command.

**Comment by sebastianas** [comment]
On 2024-10-24 07:49:13 [-0700], Tomáš Mráz wrote:
> I've created PR #25792 to fix the crashing speed command.

Why are so dismissive towards extending an existing size check?

Sebastian

**Comment by t8m** [comment]
I wanted to push you towards doing the right fix. However as I did it myself, and other assembler modules basically do the same (i.e. no-op for less than block size) I am OK with this being added as well.

**Comment by Sashan** [comment]
This PR should be close, because issue got fixed via #25792

**Comment by tom-cosgrove-arm** [comment]
Closing as fixed via #25792

**Comment by sebastianas** [comment]
On 2024-11-25 05:52:26 [-0800], Tom Cosgrove wrote:
> Closed #25637.

It makes me sad.
Could someone please explain in plain english _why_ other arch assmebly
code checks the size and it is okay but adding this to sparc gets
rejected claming that it is obsolete API and it makes no sense call it
and the testsuite should be fixed instead?

Sebastian

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to all the active branches. Thank you for your contribution.
